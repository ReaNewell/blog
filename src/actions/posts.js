import database from '../firebase/firebase';
import { firebase, storage } from '../firebase/firebase';

export const addPost = ({id, title, body, postDate, categories}) => ({
    type: "ADD_POST",
    post: {
        id,
        title,
        body,
        postDate,
        categories
    }
});
export const startAddPost = (postData = {}, image) => {
    return (dispatch) => {
        let currentPost;
        const {
            title = "",
            body = "",
            postDate = {
                day: new Date().getDate(),
                month: new Date().getMonth(),
                year: new Date().getFullYear()
            },
            categories = []
        } = postData;
        const post = { title, body, postDate, categories };

        return database.ref(`posts`).push(post).then((ref) => {
            dispatch(addPost({
                id: ref.key,
                ...post
            }));
            currentPost = ref.key;
        }).then(() => {
            if (image) {
                const fileName = image.name;
                const storageRef = storage.ref(`blogPictures/${fileName}`);
                const uploadTask = storageRef.put(image);
                let downloadURL;

                // Listen for state changes, errors, and completion of the upload.
                uploadTask.on(firebase.storage.TaskEvent.STATE_CHANGED, function(snapshot) {
                    // Get task progress, including the number of bytes uploaded and the total number of bytes to be uploaded
                    var progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
                    console.log('Upload is ' + progress + '% done');
                    switch (snapshot.state) {
                        case firebase.storage.TaskState.PAUSED: // or 'paused'
                        console.log('Upload is paused');
                        break;
                        case firebase.storage.TaskState.RUNNING: // or 'running'
                        console.log('Upload is running');
                        break;
                    }
                    }, function(error) {
    
                    }, function() {
                        // Upload completed successfully, now we can get the download URL
                        console.log('Upload complete.')
                        downloadURL = uploadTask.snapshot.downloadURL;
                        return database.ref(`posts/${currentPost}`).update({ 
                            postPicture: downloadURL,
                            postPictureName: fileName
                        })
                    });
            }
        })
    }
}

export const removePost = (id) => ({
    type: "REMOVE_POST",
    id
});
export const startRemovePost = (id, imageName) => {
    return (dispatch) => {
        return database.ref(`posts/${id}`).remove().then(() => {
            dispatch(removePost(id));
        }).then(() => {
            if (imageName) {
                const storageRef = storage.ref(`blogPictures/${imageName}`);
                storageRef.delete().then(() => {
                    console.log('Image deleted');
                }).catch((error) => {
                    console.log(error);
                });
            }
        });
    }
}

export const setPosts = (posts) => ({
    type: "SET_POSTS",
    posts
});
export const startSetPosts = () => {
    return (dispatch) => {
        return database.ref(`posts`).once('value').then((snapshot) => {
            let posts = [];

            snapshot.forEach((childSnapshot) => {
                posts.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setPosts(posts));
        });
    };
};

export const updatePost = (id, updates) => ({
    type: "UPDATE_POST",
    id,
    updates
});
export const startUpdatePost = (id, updates, currentPicture, picture) => {
    return (dispatch) => {
        return database.ref(`posts/${id}`).update({
            title: updates.title,
            body: updates.body
        }).then(() => {
            dispatch(updatePost(id, updates));
        }).then(() => {
            if (picture) {
                const fileName = picture.name;
                const storageRef = storage.ref(`/blogPictures/${fileName}`);
                const uploadTask = storageRef.put(picture);
                let downloadURL;

                // Listen for state changes, errors, and completion of the upload.
                uploadTask.on(firebase.storage.TaskEvent.STATE_CHANGED, function(snapshot) {
                // Get task progress, including the number of bytes uploaded and the total number of bytes to be uploaded
                var progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
                console.log('Upload is ' + progress + '% done');
                switch (snapshot.state) {
                    case firebase.storage.TaskState.PAUSED: // or 'paused'
                    console.log('Upload is paused');
                    break;
                    case firebase.storage.TaskState.RUNNING: // or 'running'
                    console.log('Upload is running');
                    break;
                }
                }, function(error) {

                }, function() {
                    // Upload completed successfully, now we can get the download URL
                    console.log('Upload complete.')
                    downloadURL = uploadTask.snapshot.downloadURL;
                    return database.ref(`posts/${id}`).update({ 
                        postPicture: downloadURL,
                        postPictureName: fileName
                    }).then(() => {
                        if (currentPicture && currentPicture !== fileName) {
                            return storage.ref(`/blogPictures/${currentPicture}`).delete();
                        }
                    });
                });
            }
        })
    }
}