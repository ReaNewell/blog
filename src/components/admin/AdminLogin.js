import React from 'react';

class AdminLogin extends React.Component {
    render() {
        return (
            <div>Please login, sir.</div>
        )
    }
}

export default AdminLogin;