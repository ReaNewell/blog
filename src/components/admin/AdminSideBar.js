import React from 'react';
import { Link } from 'react-router-dom';

export class AdminSideBar extends React.Component {
    chooseActiveButton = (buttonPath) => {
        return this.props.path == buttonPath ? "admin-sidebar__button--active" : "admin-sidebar__button";
    }
    render() {
        return (
            <div className="admin-sidebar">
                <h3 className="admin-sidebar__welcome">Hello, Rea</h3>
                <Link className={this.chooseActiveButton('create')} to="/admin">Create Post</Link>
                <Link className={this.chooseActiveButton('edit')} to="/admin/edit-posts">Edit Posts</Link>
                <Link className={this.chooseActiveButton('settings')} to="/admin/settings">Settings</Link>
                <button className="admin-sidebar__button--logout">Logout</button>
            </div>
        )
    }
}

export default AdminSideBar;