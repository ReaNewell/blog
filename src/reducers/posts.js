const defaultPostsReducerState = [];

export default (state = defaultPostsReducerState, action) => {
    switch (action.type) {
        case "ADD_POST":
            return state.concat({ ...action.post })
        default:
            return state;
    }
}