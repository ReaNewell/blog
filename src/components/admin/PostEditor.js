import React from 'react';
import { connect } from 'react-redux';
import { startAddPost } from '../../actions/posts';
import { init } from 'pell';
import 'pell/dist/pell.css';

class PostEditor extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            error: "",
            title: "",
            html: null
        }
    };
    componentDidMount = () => {
        this.editor = init({
            element: document.getElementById('editor'),
            onChange: html => this.setState({ html }),
            actions: ['bold', 'underline', 'italic', 'heading1', 'heading2']
        })
    }
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({ title }));
    }
    submitPost = () => {
        const image = document.getElementsByClassName('post-editor__image-input')[0].files[0];

        if (!this.state.html || !this.state.title) {
            this.setState(() => ({ error: "You must have a title and body."}));
        } else {
            this.setState(() => ({ error: ""}));
            this.props.startAddPost({
                title: this.state.title,
                body: this.state.html
            }, image);
            this.setState(() => ({ title: ""}));
            this.setState(() => ({ html: ""}));
        }
    };
    render() {
        return (
            <div className="post-editor">
                <input className="post-editor__title-input" onChange={this.onTitleChange} placeholder="Article Title" type="text"/>
                <div id="editor" className="pell" />
                <input 
                    accept='.jpg, .jpeg' 
                    className="post-editor__image-input" 
                    type='file'/>
                <button className="post-editor__button" onClick={this.submitPost}>Submit Post</button>
            </div>
        )
    };
}
const mapDispatchToProps = (dispatch) => ({
    startAddPost: (postInfo, image) => dispatch(startAddPost(postInfo, image))
});

export default connect(undefined, mapDispatchToProps)(PostEditor);