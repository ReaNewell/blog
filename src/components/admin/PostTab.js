import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { startRemovePost } from '../../actions/posts';

class PostTab extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            warning: false
        }
    }
    closeWarning = () => {
        this.setState(() => ({ warning: false }));
    }
    showWarning = () => {
        this.setState(() => ({ warning: true }));
    }
    removePost = (id, image) => {
        this.props.startRemovePost(id, image);
    }
    render() {
        return (
            <div>
                <h3>{this.props.post.title}</h3>
                <p>{this.props.post.postDate.month}/{this.props.post.postDate.day}/{this.props.post.postDate.year}</p>
                <Link to={`/admin/edit-posts/${this.props.post.title}`}>Edit Post</Link>
                {this.state.warning ? (
                    <div>
                        <p>Are you sure?</p>
                        <button onClick={this.removePost(this.props.post.id, this.props.post.postPictureName)}>
                            Yes
                        </button>
                        <button onClick={this.closeWarning}>No</button>
                    </div>
                ) : (
                    <button onClick={this.showWarning}>Delete Post</button>
                )}
            </div>
        )
    }
}
const mapDispatchToProps = (dispatch) => ({
    startRemovePost: (id, image) => dispatch(startRemovePost(id, image))
})

export default connect(undefined, mapDispatchToProps)(PostTab);