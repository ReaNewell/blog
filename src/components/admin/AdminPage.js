import React from 'react';
import AdminSideBar from './AdminSideBar';

class AdminPage extends React.Component {
    render() {
        return (
            <div>
                <AdminSideBar />
                <div>
                    <h1>Admin Page</h1>
                </div>
            </div>
        )
    }
}

export default AdminPage;