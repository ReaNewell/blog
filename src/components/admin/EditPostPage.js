import React from 'react';
import { connect } from 'react-redux';
import AdminSideBar from './AdminSideBar';
import PostEditor from './PostEditor';
import PostTab from './PostTab';

export class EditPostPage extends React.Component {
    render() {
        return (
            <div className="dashboard">
                <AdminSideBar path="edit"/>
                { this.props.currentPost ? (
                    <div>
                        <h1>Edit Posts</h1>
                        <h2>Make changes to article, then save.</h2>
                        <PostEditor 
                            history = {this.props.history} 
                            post={this.props.currentPost}
                            updating={true}
                        />
                    </div>
                ) : (
                    <div className="admin-page">
                        <h1 className="admin-page__title">Edit Posts</h1>
                        <h2 className="admin-page__subtitle">Posted Articles</h2>
                        <div>
                            { this.props.posts.map( post => (
                                <PostTab 
                                    key={post.id}
                                    post={post}
                                />
                            ))}
                        </div>
                    </div>
                )}
            </div>
        )
    }
}
const mapStateToProps = (state, props) => ({
    currentPost: props.match.params.title ? state.posts.find((post) => post.title === props.match.params.title.replace('-', ' ')) : null,
    posts: state.posts
})

export default connect(mapStateToProps)(EditPostPage);