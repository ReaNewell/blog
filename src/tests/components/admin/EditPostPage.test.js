import React from 'react';
import { shallow } from 'enzyme';
import { EditPostPage } from '../../../components/admin/EditPostPage';
import posts from '../../fixtures/posts';

test('should render EditPostPage component correctly when current post is selected', () => {
  const wrapper = shallow(<EditPostPage posts={posts} currentPost={posts[1]} />);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('h2').at(0).text()).toBe('Make changes to article, then save.');
});

test('should render EditPostPage component correctly when current post is not selected', () => {
  const wrapper = shallow(<EditPostPage posts={posts} />);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('h2').at(0).text()).toBe('Posted Articles');
});