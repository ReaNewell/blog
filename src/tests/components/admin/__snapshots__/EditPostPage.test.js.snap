// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditPostPage component correctly when current post is not selected 1`] = `
<div
  className="dashboard"
>
  <AdminSideBar
    path="edit"
  />
  <div
    className="admin-page"
  >
    <h1
      className="admin-page__title"
    >
      Edit Posts
    </h1>
    <h2
      className="admin-page__subtitle"
    >
      Posted Articles
    </h2>
    <div>
      <Connect(PostTab)
        key="123"
        post={
          Object {
            "body": "This is the first blog post.",
            "categories": Array [
              "css",
              "html",
            ],
            "id": "123",
            "postDate": Object {
              "day": 1,
              "month": 2,
              "year": 2017,
            },
            "postPicture": "postPicture",
            "postPictureName": "postPictureName",
            "title": "Title One",
          }
        }
      />
      <Connect(PostTab)
        key="221"
        post={
          Object {
            "body": "This is the second blog post.",
            "categories": Array [
              "javascript",
              "html",
            ],
            "id": "221",
            "postDate": Object {
              "day": 1,
              "month": 4,
              "year": 2018,
            },
            "postPicture": "postPictureNumberTwo",
            "postPictureName": "postPictureNameNumberTwo",
            "title": "Title Two",
          }
        }
      />
      <Connect(PostTab)
        key="421"
        post={
          Object {
            "body": "This is the third blog post.",
            "categories": Array [
              "react",
              "javascript",
            ],
            "id": "421",
            "postDate": Object {
              "day": 5,
              "month": 9,
              "year": 2018,
            },
            "postPicture": "postPictureNumberThree",
            "postPictureName": "postPictureNameNumberThree",
            "title": "Title Three",
          }
        }
      />
      <Connect(PostTab)
        key="522"
        post={
          Object {
            "body": "This is the fourth blog post.",
            "categories": Array [
              "ruby",
              "python",
            ],
            "id": "522",
            "postDate": Object {
              "day": 27,
              "month": 3,
              "year": 2017,
            },
            "postPicture": "postPictureNumberFour",
            "postPictureName": "postPictureNameNumberFour",
            "title": "Title Four",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`should render EditPostPage component correctly when current post is selected 1`] = `
<div
  className="dashboard"
>
  <AdminSideBar
    path="edit"
  />
  <div>
    <h1>
      Edit Posts
    </h1>
    <h2>
      Make changes to article, then save.
    </h2>
    <Connect(PostEditor)
      post={
        Object {
          "body": "This is the second blog post.",
          "categories": Array [
            "javascript",
            "html",
          ],
          "id": "221",
          "postDate": Object {
            "day": 1,
            "month": 4,
            "year": 2018,
          },
          "postPicture": "postPictureNumberTwo",
          "postPictureName": "postPictureNameNumberTwo",
          "title": "Title Two",
        }
      }
      updating={true}
    />
  </div>
</div>
`;
